numpy/core/tests/test_function_base.py

```diff
@@ -407,3 +407,67 @@ def test_round_negative(self):
         y = linspace(-1, 3, num=8, dtype=int)
         t = array([-1, -1, 0, 0, 1, 1, 2, 3], dtype=int)
         assert_array_equal(y, t)
+
+    def test_datetime(self):
+        t1 = datetime64("2018-12-12 06:00:00")
+        t2 = datetime64("2018-12-12 07:00:00")
+        quarters = array([0, 15*60, 30*60, 45*60, 60*60], dtype="m8[s]")
+        assert_array_equal(linspace(t1, t2, 5), t1 + quarters)
+
+    def test_timedelta(self):
+        t1 = timedelta64(0, "s")
+        t2 = timedelta64(3600, "s")
+        quarters = array([0, 15*60, 30*60, 45*60, 60*60], dtype="m8[s]")
+        assert_array_equal(linspace(t1, t2, 5), t1 + quarters)
+
+    def test_nat(self):
+        t1 = datetime64("NaT")
+        t2 = datetime64("2020-01-01")
+        actual = linspace(t1, t2, 5)
+        expected = array([t1]*4+[t2])
+        assert_array_equal(actual, expected)
+        actual = linspace(t1, t2, 5, False)
+        expected = array([t1]*5)
+        assert_array_equal(actual, expected)
+
+    def test_datetime_mixedunits(self):
+        t1 = datetime64("2020-01-01").astype("M8[D]")
+        t2 = datetime64("2020-02-01").astype("M8[s]")
+        actual = linspace(t1, t2, 31, False, dtype="M8[ms]")
+        expected = array(
+            [datetime64(f"2020-01-{i:0>2}") for i in range(1, 32)], "M8[ms]")
+        assert_array_equal(actual, expected)
+
+    def test_datetime_div0(self):
+        t1 = datetime64("2020-01-01")
+        t2 = datetime64("2020-01-02")
+        actual, step = linspace(t1, t2, 1, True, True, "M8[D]")
+        expected = array(t1)
+        assert_array_equal(step, array(datetime64("NaT")))
+        assert_array_equal(actual, expected)
+
+    def test_datetime_multidimensional(self):
+        t1 = datetime64("2020-07-01")
+        t2 = datetime64("2020-08-01")
+        t3 = datetime64("2020-09-01")
+        actual = linspace([t1, t2], [t2, t3], 31, False, dtype="M8[D]")
+        expected0 = array(
+            [datetime64(f"2020-07-{i:0>2}") for i in range(1, 32)])
+        expected1 = array(
+            [datetime64(f"2020-08-{i:0>2}") for i in range(1, 32)])
+        expected = array([expected0, expected1]).T
+        assert_array_equal(actual, expected)
+
+    def test_datetime_nicely_spaced_step(self):
+        t1 = datetime64("2020-01-01T00:00:00")
+        t2 = datetime64("2020-01-01T00:00:01")
+        actual = diff(linspace(t1, t2, 33, dtype="M8[ms]"))
+        expected = diff(linspace(0, 1000, 33, dtype="int64"))
+        assert_array_equal(actual, expected)
```

Comment:
I wonder what the step is in this case, since we cannot represent it as a timedelta64?

Reply:
The step size is knowable via `linspace(..., retstep=True)` and in this case it is 31.25 which is the same as the integer case.

Reply:
But that means we return a floating point step, which is not compatible with the original array (you cannot add float to datetime64, and even if you do it may use the wrong unit).  I suppose a strange solution could be to not support `retstep` so that we can't return an arguably incorrect value (not that I like that solution).

Reply:
Ok, how about cast to the internal `delta_dt` which in this case results in a step size of `timedelta64(31,'ms')` instead of 31.25?